### PR TITLE
Speed up source deletion

### PR DESCRIPTION
**Pull request by creviera**
# Description

Fixes https://github.com/freedomofpress/securedrop-client/issues/1344
Fixes https://github.com/freedomofpress/securedrop-client/issues/1343

# Test Plan

- [ ] deleting a source now happens right away (no waiting until a sync)
- [ ] follow STR in https://github.com/freedomofpress/securedrop-client/issues/1343 and confirm fix

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [ ] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [ ] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [ ] No database schema changes are needed

### CODE REVIEW

File: tests/gui/test_widgets.py [hunk 1]
```diff
@@ -950,6 +950,70 @@ def test_MainView_set_conversation(mocker):
...
+    mv.source_conversations = {}
+    mv.source_conversations["123"] = conversation_wrapper
+    mv.source_list.controller = mocker.MagicMock()
+    mv.source_list.update([source])
+    assert not mv.source_list.source_items == {}
+
+    mv._delete_source("123")
+
+    assert mv.source_conversations == {}
+    assert mv.source_list.source_items == {}
```
Comment (gonzalo-bulnes) [inline tests/gui/test_widgets.py:969]: Python question / maybe nitpicking: is the intention to check that the collections are not empty, then empty? And could a `len` check work/convey that better?

I'm thinking something in the lines of:

```diff
- assert not mv.source_list.source_items == {}
+ assert len(mv.source_list.source_items) > 0

- assert mv.source_list.source_items == {}
+ assert len(mv.source_list.source_items) == 0

File: tests/gui/test_widgets.py [hunk 2]
```diff
@@ -950,6 +950,70 @@ def test_MainView_set_conversation(mocker):
...
+    mv.source_conversations["123"] = SourceConversationWrapper(source2, mocker.MagicMock())
+    mv.source_list.controller = mocker.MagicMock()
+    mv.source_list.update([source1, source2])
+    assert not mv.source_list.source_items == {}
+
+    mv._delete_source("123")
+    mv._delete_source("abc")
+
+    assert mv.source_conversations == {}
+    assert mv.source_list.source_items == {}
```
Comment (gonzalo-bulnes) [inline tests/gui/test_widgets.py:990]: I like how these tests use mocks to simplify the setup but test the outcome of the deletions! (i.e. don't mock the object under test) :star:

File: tests/test_sync.py [hunk 3]
```diff
@@ -158,12 +158,14 @@ def test_ApiSync_on_sync_failure(mocker, session_maker, homedir):
...
     api_sync = ApiSync(mocker.MagicMock(), session_maker, mocker.MagicMock(), homedir)
     sync_failure = mocker.patch.object(api_sync, "sync_failure")
+    singleShot_fn = mocker.patch("securedrop_client.sync.QTimer.singleShot")
 
     error = Exception()
 
     api_sync.on_sync_failure(error)
 
     sync_failure.emit.assert_called_once_with(error)
+    singleShot_fn.assert_called_once_with(15000, api_sync.api_sync_bg_task.sync)
```
Comment (gonzalo-bulnes) [inline tests/test_sync.py:168]: In relation to my comment above: testing time-related things is difficult, so this seems like an OK compromise to me even though it tests the implementation more than the outcome.

**Review by gonzalo-bulnes — COMMENTED** [review]
Looks good to me, but I'm not confident I understand the sync-ing behavior well enough to give a full green tick. : )

**Comment by creviera** [comment]
I was going to add the same timestamp logic as used for conversation deletion, but i don't think we need it. Calling on @eloquence to try to break i! If it's good as is, I'll remove the timestamp from the new deletion signal. If the "ghost" issue happens, I'll add logic using the timestamp.

File: tests/gui/test_widgets.py (hunk 1 above)
Comment (creviera) [reply tests/gui/test_widgets.py:969]: i agree, that's cleaner

File: tests/test_sync.py (hunk 3 above)
Comment (creviera) [reply tests/test_sync.py:168]: yeah, i went back and forth once on this, but figured we'd want to make sure this "business rule" of 15 seconds is confirmed.

**Comment by creviera** [comment]
@gonzalo-bulnes feel free to try to break this too plz X). I didn't see your review until after I posted that, but always appreciate the way erik finds user-facing issues in my code!

File: tests/test_sync.py (hunk 3 above)
Comment (gonzalo-bulnes) [reply tests/test_sync.py:168]: I agree it's worth having some guarantees at least, and this code is unlikely to become high-traffic (as in change a lot), so there is less downside to testing its implementation directly. :+1: 

My last attempt at testing time-dependent changes resulted in https://github.com/gonzalo-bulnes/stime :wink:

**Comment by gonzalo-bulnes** [comment]
> Revert "Introduce immediate sync upon source/conversation deletion" 

:bulb: So this means using a single-shot timer trigger is an known-good solution, am I correct?

**Comment by eloquence** [comment]
I've taken this for a spin today and overall it looks great! Deletion is indeed much faster, and I'm not able to reproduce #1343 any longer.

I do see an issue with sources sometimes re-appearing until the next sync (similar to #858). These "ghost sources" can be interacted with, but if you attempt to reply to them, you get a "Failed to send" because they've already been deleted on the sever. The error is logged as "bad request" on the client.

The way I'm able to reproduce it:

1. Submit a bunch of sources (can be just messages)
2. Delete all of them in quick succession in the client
3. Observe the source list carefully

What happens is that sometimes a source like `curved jerky` (actual name I got today :P) will be removed, then briefly reappear in the source list, then be removed after the next sync completes successfully.

**Comment by creviera** [comment]
> The way I'm able to reproduce it:
> 
> 1. Submit a bunch of sources (can be just messages)
> 2. Delete all of them in quick succession in the client
> 3. Observe the source list carefully

Yup, confirmed. I think I know how to fix... I'll keep the fix in a separate commit

**Comment by eloquence** [comment]
Taking this for another quick spin now.

**Comment by eloquence** [comment]
Functionally all looks great to me:
- Source deletion is performed quickly (#1344)
- Animation is timed correctly (#1343)
- No "ghost" sources re-appear due to stale sync data
- Files are removed from disk as expected
- Records are removed from database as expected

**Review by conorsch — APPROVED** [review]
Works great! The removal is snappy and complete. Tested several times against an onion service, and it was pleasant responsive. Also tested a pathological edge where the server _doesn't_ reply with a 200, and the client appropriately raised the error and the source was _not_ removed from the UI prematurely:

failed-to-delete-at-server

If the server did actually delete the source in question, despite responding not-ok, then the subsequent sync dropped the source from the UI, which is a nice response. When I edited the server code to respond not-ok and preserve the source in the server db, subsequent client syncs did not remove the source from the UI—also good!

I tacked on one small cleanup commit. LGTM.

**Comment by conorsch** [comment]
> I tacked on one small cleanup commit.

Ah, was already handled in https://github.com/freedomofpress/securedrop-client/commit /d9e5546e9f5ce1cb11dbaba0049e70d683eec1c7 ; just needed a rebase here.

**Review by conorsch — APPROVED** [review]
